3rdParty/iresearch/core/formats/columnstore2.cpp

```diff
@@ -801,82 +909,176 @@ doc_iterator::ptr fixed_length_column::iterator(bool /*consolidation*/) const {
-      std::optional<std::string>&& name,
-      bstring&& payload,
-      column_header&& hdr,
-      column_index&& index,
-      index_input& index_in,
-      const index_input& data_in,
-      compression::decompressor::ptr&& inflater,
-      encryption::stream* cipher) {
+  template<typename ColumnType>
+  static column_ptr read(std::optional<std::string>&& name, bstring&& payload,
+                         column_header&& hdr, column_index&& index,
+                         index_input& index_in, const index_input& data_in,
+                         compression::decompressor::ptr&& inflater,
+                         encryption::stream* cipher) {
     auto blocks = read_blocks_sparse(hdr, index_in);
-
-    return memory::make_unique<sparse_column>(
-        std::move(name), std::move(payload),
-        std::move(hdr), std::move(index),
-        data_in, std::move(inflater), cipher,
-        std::move(blocks));
-  }
-
-  sparse_column(
-      std::optional<std::string>&& name,
-      bstring&& payload,
-      column_header&& hdr,
-      column_index&& index,
-      const index_input& data_in,
-      compression::decompressor::ptr&& inflater,
-      encryption::stream* cipher,
-      std::vector<column_block>&& blocks)
-    : column_base{ std::move(name), std::move(payload),
-                   std::move(hdr), std::move(index),
-                   data_in, cipher },
-      blocks_{std::move(blocks)},
-      inflater_{std::move(inflater)} {
+    return memory::make_unique<ColumnType>(
+        std::move(name), std::move(payload), std::move(hdr), std::move(index),
+        data_in, std::move(inflater), cipher, std::move(blocks));
+  }
+
+  sparse_column(std::optional<std::string>&& name, bstring&& payload,
+                column_header&& hdr, column_index&& index,
+                const index_input& data_in,
+                compression::decompressor::ptr&& inflater,
+                encryption::stream* cipher, std::vector<column_block>&& blocks)
+      : column_base{std::move(name), std::move(payload),
+                    std::move(hdr),  std::move(index),
+                    data_in,         cipher},
+        blocks_{std::move(blocks)},
+        inflater_{std::move(inflater)} {
     assert(header().docs_count);
     assert(ColumnType::kSparse == header().type);
   }
 
   virtual doc_iterator::ptr iterator(bool /*consolidation*/) const override;
 
+  void make_buffered(
+      irs::index_input& in, const memory_accounting_f& memory_accounter,
+      range<std::unique_ptr<column_reader>> next_sorted_columns) override {
+    buffered_memory_accounter_ = memory_accounter;
+    auto& hdr = mutable_header();
+    if (!is_encrypted(hdr)) {
+      if (make_buffered_data<false>(hdr, in, blocks_, column_data_,
+                                    next_sorted_columns, nullptr)) {
+        buffered_input_ = std::make_unique<irs::bytes_ref_input>(
+            irs::bytes_ref{column_data_.data(), column_data_.size()});
+      }
```

Comment:
Is there a problem if `make_buffered_data` succeeds, but the following call to `make_unique` throws?

Reply:
No, allocated `column_data` would be freed by dtor. Column is in the undefined state. But index segment is "failed to open" anyway. So columns should not be used anyway.